plugins/blueprint.cpp

```diff
@@ -1517,35 +1517,30 @@ static bool do_transform(color_ostream &out,
         return false;
 
     const bool pretty = opts.format != "minimal";
-    const int32_t z_inc = start.z < end.z ? 1 : -1;
-    for (int32_t z = start.z; z != end.z; z += z_inc) {
-        for (int32_t y = start.y; y < end.y; y++) {
-            for (int32_t x = start.x; x < end.x; x++) {
-                df::coord pos(x, y, z);
-                tile_context ctx;
-                ctx.pretty = pretty;
-                for (blueprint_processor &processor : processors) {
-                    ctx.processor = &processor;
-                    if (processor.init_ctx)
-                        processor.init_ctx(pos, ctx);
-                    const char *tile_str = processor.get_tile(pos, ctx);
-                    if (tile_str) {
-                        // ensure our z-index is in the order we want to write
-                        auto area = processor.mapdata.emplace(abs(z - start.z),
-                                                              EMPTY_AREA);
-                        auto row = area.first->second.emplace(y - start.y,
-                                                              EMPTY_ROW);
-                        auto &tiles = row.first->second;
-                        if (row.second)
-                            tiles.resize(opts.width);
-                        tiles[x - start.x] = tile_str;
-                    }
-                }
+    df::coord min_pos(bounds.x_min, bounds.y_min, bounds.z_min);
+
+    bounds.forCoord([&](df::coord pos) {
+        tile_context ctx;
+        ctx.pretty = pretty;
+        for (blueprint_processor &processor : processors) {
+            ctx.processor = &processor;
+            if (processor.init_ctx)
+                processor.init_ctx(pos, ctx);
+            const char *tile_str = processor.get_tile(pos, ctx);
+            if (tile_str) {
+                auto rel_pos = pos - min_pos;
+                auto area = processor.mapdata.emplace(rel_pos.z, EMPTY_AREA);
```

Comment:
It could end up being entirely irrelevant in comparison to file I/O.

Reply:
it isn't a performance concern. it's a functionality concern. it matters whether a blueprint is specified from the top down or the bottom up because it changes where the player will have to click to apply the blueprint.

Reply:
from testing, this PR produces incorrect blueprints for both the up-down case and the down-up case.

Try these commands to test (at an appropriate cursor position):

```
blueprint 9 9 -2 updown --cursor=80,50,126
blueprint 9 9 2 downup --cursor=80,50,125
```

Reply:
Then do we want to implement the reverse `forCoord` iterator, try to fix the rest of the blueprint code, or revert the change?

Reply:
I think just revert the forCoord part of the change. I think forCoord handles the common case and special cases like this can manage it itself. We can still use `cuboid` to hold the bounds.